File: dojo20100714 - espiral - php/espiral.php
<?php

function formatar_em_espiral($matriz){
    $espiral = "";
    foreach($matriz as $linha){
        foreach($linha as $celula){
            $espiral .= $celula." ";
        }
        $espiral = rtrim($espiral). "\n";
    }
    return substr($espiral, 0, -1);
}
function direita($espiral, $coluna, $linha, $coluna_maxima, $elemento){
$espiral [$linha][$coluna]= ++$elemento;
}

function gerar_espiral($tamanho){

    $espiral = array();

    if($tamanho == 1){
            $elemento=0;
            $linha = 0;
            $coluna = 0;

            $linha_maxima = sqrt($tamanho) - 1;
            $coluna_maxima = $linha_maxima;
            $linha_minima = 0;
            $coluna_minima = 0;

            direita(&$espiral, $coluna, $linha, $coluna_maxima, $elemento);
  //          baixo(&$espiral, $linha_minima, $linha_maxima, $coluna, $elemento_inicial);
   //         esquerda();
     //       cima();

            //$espiral[$linha][$coluna] = $elemento;
    }elseif($tamanho == 4){
        $espiral[] = array(1,2);
        $espiral[] = array(4,3);
    }elseif($tamanho == 9){
        $espiral[] = array(1, 2, 3);
        $espiral[] = array(8, 9, 4);
        $espiral[] = array(7, 6, 5);
    }


    $espiral = formatar_em_espiral($espiral);

    return $espiral;
}
?>

File: 2011/20110105 - espiral - ruby/espiral.rb
class Espiral
  def initialize(linha, coluna)
    @linha = linha
    @coluna = coluna
  end

  def matriz()
    [(1..@coluna).to_a]
  end

  def coluna(indice)
    if indice == 1
      (1..@linha).to_a
    elsif indice == 2

    else
      (@coluna..@linha+@coluna-1).to_a
    end
  end

  def linha(indice)
    if indice == 1
      (1..@coluna).to_a
    elsif indice == 2
      (@coluna+1..2*@coluna).to_a.reverse
    else
      [4,3]
    end
  end
end

